### PR TITLE
Disallow empty version string

### LINKED ISSUE #181: Drop support for using empty string to get the latest S3 object version
As of https://github.com/JuliaCloud/AWSS3.jl/pull/161 we should be using `nothing` when trying to retrieve the latest version of an S3 object. In the past we supported using empty string which should be deprecated.

### PR DESCRIPTION
**Pull request by christopher-dG**
Closes #181, I think.

Alternatively, we could make this a deprecation since I don't think we did that in 0.8.x.

### CODE REVIEW

**Comment by christopher-dG** [comment]
bors try

**Comment by christopher-dG** [comment]
bors try

**Review by mattBrzezinski — APPROVED** [review]

File: test/awss3.jl [hunk 1]
```diff
@@ -205,4 +205,13 @@ function awss3_tests(config)
             @test_throws AWS.AWSException s3_delete_bucket(config, bucket_name)
         end
     end
+
+    @testset "Empty version string" begin
```
Comment (mattBrzezinski) [inline test/awss3.jl:209]: Could you add in a test for `S3Path` w/ an empty version here too?

**Comment by omus** [comment]
> Alternatively, we could make this a deprecation since I don't think we did that in 0.8.x.

The nice thing to do IMO is we deprecate this nicely in a 0.8.9 backport and drop support for `""` in 0.9

File: test/s3path.jl [hunk 2]
```diff
@@ -535,8 +537,7 @@ function s3path_tests(config)
             str_v1 = string(versioned_path_v1)
             roundtripped_v1 = S3Path(str_v1; config=config)
             @test isequal(versioned_path_v1, roundtripped_v1)
-            @test str_v1 ==
-                  "s3://" * bucket_name * "/" * key_version_file * "?versionId=" * v1
+            @test str_v1 == "s3://$(bucket_name)/$(key_version_file)?versionId=$(v1)"
```
Comment (omus) [inline test/s3path.jl:540]: May want a test where we do `"s3://...?versionId="`

**Comment by omus** [comment]
@christopher-dG is on vacation so I'll try to get this PR moving. I'll be trying to proceed with dropping support for treating `""` as `nothing` here and have a follow up PR which deprecates the behaviour for a 0.8 backport.

**Comment by omus** [comment]
bors try

**Comment by omus** [comment]
I was trying to determine the syntax of the "versionId" string and couldn't find any documentation. So I pulled down all the versions from a bucket and here were the results:
```julia
julia> lines = readlines("versions.txt")  # Extracted "VersionId" fields from `aws s3api list-object-versions`

julia> length(lines)
599220

julia> mapreduce(length, union, lines)
1-element Vector{Int64}:
 32

julia> String(sort!(reduce(union, lines)))
".0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz"
```
Based upon these results we could validate a proper "versionId" with the regex `r"^[0-9a-zA-Z\._]{32}$"`

**Comment by omus** [comment]
bors try

**Comment by omus** [comment]
Okay, so I discovered something that may change our approach here. Passing in `"versionId" = ""` is supported by the underlying AWS S3 API calls and returns the latest version of the file. The reason this is important is that I don't think we should go out of our way to check for `""` and fail when we find it. Instead, we should recommend using `nothing` and if someone uses `""` it can just rely on the API behaviour.

If however we do validate the versionId using the regex mentioned above then I think we would just disable using `""`

**Comment by omus** [comment]
bors try

**Comment by omus** [comment]
bors try

**Comment by omus** [comment]
bors try

**Comment by omus** [comment]
bors try

**Comment by omus** [comment]
bors try

**Comment by omus** [comment]
Ready for review. I made a backport PR of this here: https://github.com/JuliaCloud/AWSS3.jl/pull/202

**Comment by omus** [comment]
bors try

**Comment by omus** [comment]
bors try

**Comment by omus** [comment]
Failure looks unrelated to me.

**Comment by omus** [comment]
Rebased to resolve conflicts

**Comment by omus** [comment]
bors try

**Review by mattBrzezinski — APPROVED** [review]

**Comment by mattBrzezinski** [comment]
We really need to find a better solution to CI testing here. This is quite annoying, lets try again!

bors r+

**Review by christopher-dG — COMMENTED** [review]
Sorry for late review, but I've just looked over your changes and all is good 👍
